M2 init project

Demitsuru opened the pull request.

Commit:
init

Commit:
dsl practice

Commit:
added routing and base test

Commit:
added routing for creation of customers; http sample file; dependencies

Commit:
initial commit for common and multiplatform project

svok commented on backend-common/build.gradle.kts:
```diff
@@ -0,0 +1,51 @@
+version = rootProject.version
+
+val ktorVersion: String by project
+val kotestVersion: String by project
```
поскольку эти переменные используются только в зависимостях, рекомендую ограничить их область видимости и поместить в блок dependencies

Demitsuru replied:
благодарю. подправил

svok commented on backend-common/src/main/kotlin/builders/marketplace/dsl/UserConfig.kt:
```diff
@@ -0,0 +1,73 @@
+                }
+    }
+
+    fun address(block: AddressConfig.() -> Unit) {
```
Очень хорошая задумка. Еще чуть-чуть упростить код можно, если использовать сокращенную запись функции. В выбранном подходе она очень подойдет

Demitsuru replied:
благодарю. тоже подправил

svok commented on backend-common/src/main/kotlin/builders/marketplace/models/Address.kt:
```diff
@@ -0,0 +1,9 @@
+
+data class Address(
+        val postCode: String = "",
+        val firstLine: String = "",
```
У вас разные модели в разном стиле: какие-то мутабельные, какие-то иммутабельные. Рекомендую определиться и выбрать один стиль.

Если мы говорим о внутренних моделях, то чаще всего они мутабельные, поскольку нам с ними приходится постоянно работать. Если будет много многопоточности, тогда можно подумать об иммутабельности

Demitsuru replied:
исправил

svok commented on build.gradle.kts:
```diff
@@ -0,0 +1,24 @@
+    }
+
+    tasks.withType<KotlinCompile> {
+        kotlinOptions.jvmTarget = "1.8"
```
Рекомендую 11. Мы будем ориентироваться на нее. 8 ява уж очень устарела

Demitsuru replied:
исправил. по привычке добавил 8

svok commented on build.gradle.kts:
```diff
@@ -0,0 +1,24 @@
+    }
+
+    tasks.withType<Wrapper> {
+        gradleVersion = "6.6.1"
```
Версию грэдла проще указать в rootProject.projectDir/gradle/wrapper/gradle-wrapper.properties.
Меньше замусоривается build.gradle.kts

Там, кстати уже указано. Этот блок можно удалить

Demitsuru replied:
исправил

svok commented on settings.gradle.kts:
```diff
@@ -0,0 +1,31 @@
+}
+
+include("backend-common")
+include("multiplatform-common")
```
Я предпочитаю названия подпроектам давать полные, включающие префикс проекта. Да, мавен-репозитарий добавит groupId и все будет понятно. Но проблемы возникают, когда мы используем другие плагины: nodejs, docker и пр. У них нет groupId, а настраивать каждый раз имя бывает утомительно

Demitsuru replied:
исправил

svok reviewed (commented):
Все хорошо получилось, дополнительно ценю эксперименты с ktor-ом, расширениями и DSL. Рекомендации оставил в комментариях.

Commit:
refactoring:
- renamed modules;
- changed scopes in build.gradle.kts;
- changed models to immutables, added lateinit to dsl configs
- removed wildcard imports;
